crates/ra_ide/src/completion/complete_trait_impl.rs

```diff
@@ -101,6 +101,36 @@ pub(crate) fn complete_trait_impl(acc: &mut Completions, ctx: &CompletionContext
     }
 }
 
+fn completion_match(ctx: &CompletionContext) -> Option<(SyntaxNode, ImplDef)> {
+    let (trigger_idx, trigger) =
+        ctx.token.ancestors().enumerate().find(|(_idx, p)| match p.kind() {
+            SyntaxKind::FN_DEF
+            | SyntaxKind::TYPE_ALIAS_DEF
+            | SyntaxKind::CONST_DEF
+            | SyntaxKind::NAME_REF
+            | SyntaxKind::BLOCK_EXPR => true,
+            _ => false,
+        })?;
+    let (impl_def_idx, impl_def) =
+        ctx.token.ancestors().enumerate().skip(trigger_idx + 1).find_map(|(idx, p)| {
+            match p.kind() {
+                SyntaxKind::IMPL_DEF => ast::ImplDef::cast(p).map(|p| (idx, p)),
+                _ => None,
+            }
+        })?;
+    let _is_nested = ctx
+        .token
+        .ancestors()
+        .skip(trigger_idx + 1)
+        .take(impl_def_idx - trigger_idx - 1)
+        .find_map(|p| match p.kind() {
+            SyntaxKind::FN_DEF | SyntaxKind::BLOCK => Some(()),
```

Comment:
I think this check has the wrong "modality": it tries to black-list bad cases, instead of explicitly allowing good cases. There are usually far more (unexpected) bad cases, so white lists tend to work better in similar situation.  Specifically, this I think won't work for

```rust
impl Foo for Bar {
  const C: &'static [()] = [(); cursor_here]
}
```

I would write a very dumb check here, like `If the token is IDENT, make sure that nth parent is IMPL_ITEM_LIST`, and then repeat for various kinds of tokens. I would probably avoid calling ancestors at all.

Reply:
Switched to explicit offsets and removed the black-list approach.